### ISSUE TITLE
[C#] Client channel socket leaks unless read stream drained explicitly

### ISSUE THREAD

**Issue by kkm000** [grpc/grpc#8451]
This is still happening in v1.0.0, and I suspect has always been there.

I am using bidirectional streaming RPCs to talk to server. The server sends an exactly known number of messages back. Therefore  I am not reading the response stream to EOF, but rather end the call as soon as I get the data on the client.

If the call is disposed by the client without first reading the server stream to EOF, channel will not close its socket on shutdown, and a warning message is reported: _W1020 18:40:55.407708 Grpc.Core.Channel Channel shutdown was called but there are still 1 active calls for that channel._

In my debugging case, the server returns exactly one message. I can trace the problem as far as this. The socket handle is locked via a call to `ChannelSafeHandle.DangerousAddRef()` without a matching release on call end. `AsyncCallBase.ReleaseResourcesIfPossible()` fails to call ReleaseResources() when called from `AsyncCall.HandleFinished()`, because the member flag `readingDone` is never set to `true` in this scenario.

I do not understand the logic of the code to tell if simply setting `readingDone = true` before [the call to `ReleaseResourcesIfPossible()` from `HandleFinished()`](https://github.com/grpc/grpc/blob/v1.0.0/src/csharp/Grpc.Core/Internal/AsyncCall.cs#L498) is all that is missing. Maybe also creating a `streamingReadTcs` is required to complete possible pending reads? Is the call actually cancelled (in the gRPC sense) when `Dispose()` is called?

I work around the bug by calling `MoveNext` on the calls' reader in a loop until it returns `false`, which, in my case, it immediately does. But this a less than fantastic workaround.

**Comment by jtattermusch** [comment]
Do you have a code snippet that replicates this?

What platform are you on? Also, from your description it seems it's about a call on the client side, is that right?

Btw, for testing various orderings in which operation completions can arrive to a call, we have this test:
https://github.com/grpc/grpc/blob/9f4e828343a78a24988e638cf7ac865cf3947cbe/src/csharp/Grpc.Core.Tests/Internal/AsyncCallTest.cs  (it runs the call handling logic against a mocked-out C core API and verifies that the behavior is as expected).
Would you be able to express your scenario as a test case there?

**Comment by jtattermusch** [comment]
To understand this correctly, you basically read X (a known number) of responses on the client side and then you stop reading because you know that's all the messages that will be sent from the server?

I think if you are expecting X responses, it is a good practice to make sure that's the end of the stream and that there hasn't been an error:
```
// read X msgs here

GrpcPreconditions.CheckState(! await responseStream.MoveNext());  // check this is end of the stream.
```

Alternatively, you could use an extensions method from Grpc.Core.Utils and do
````
var responses = await responseStream.ToList();   // read all the responses
GrpcPreconditions(responses.Count == X); // make sure we got the expected num of responses
```
in which case you don't need to invoke any extra MoveNext() yourself.

**Comment by kkm000** [comment]
> Do you have a code snippet that replicates this?

I'll make one tonight, based on one of the examples. Should be easy.

> What platform are you on? Also, from your description it seems it's about a call on the client side, is that right?

Client is a 32-bit process on Windows. Yes, the bug I see happens on the client side.

>Btw, for testing various orderings in which operation completions can arrive to a call, we have this test

I do not see it verifying that a channel indeed has the call counter of 0 in this test. If possible, that should be tested in the `Assert*` methods.

Let me see if I can repro in the test.

> To understand this correctly, you basically read X (a known number) of responses on the client side and then you stop reading because you know that's all the messages that will be sent from the server?

Oh believe me, this is complex. Both endpoints have the message flow that depends on previous server responses. 

```
Client: I have this data for you.
Server: (1st message) Thanks, but I do not need it.
Client (disposes the call without reading to EOF).
```

```
Client: I have this data for you.
Server: (1st message) Please send!
Client: message, message, ..., EOF
Server churns for a while then (2nd message): Here are your results:...
Client (disposes the call without reading to EOF).
```

The second service has an even more complex protocol, where the server emits messages both in response to a flag in a client's message (at most once per client message, 0 to many per call), and a final response upon the end of client's send stream. This happens in real world time, client messages encapsulate real time data. Upon receiving the next reply from the server, the client can decide it has a good enough response and dispose of the call. The server (C++) sees the call as canceled just fine; but the C# channel has its call count permanently increased after the cancellation.

**Comment by kkm000** [comment]
I reproduced the bug in the MR mentioned above. Simply break out of the `while(ReadNext())` loop to have a channel socket to never be released even after channel shutdown. Exactly what I see in our production. I'll try to build a test for it.

Since my setup trims EOL whitespace, it's easier to look at the changes with `?w=1`: https://github.com/grpc/grpc/pull/8471/files?w=1

**Comment by kkm000** [comment]
I wrote 2 unit tests that reproduced the problem; the single-line change in AsyncCallBase.cs fixed it and did not break anything else. Please check if the fix makes sense to you, and let's wait for the integration tests to return. Please do not commit yet even if LG, I am going to give the change a run in our QA milieu.

**Comment by jtattermusch** [comment]
I think this might be still an issue.

**Comment by dadude999** [comment]
I still see this behavior as well. My code is essentially [this](https://gist.github.com/dadude999/ac93f6d70d281b673356e112d4c9f4d4). Like @kkm000, I currently work around the problem by changing the `break` to a `continue` such that the stream is always read to the end. But it would be nice not to have to do that, because this occurs in the context of shutting down an ASP.NET Core server, which is a time-sensitive operation, and at that point, I don't care about any pending messages that might be queued.

**Comment by tonydnewell** [comment]
I can confirm that the socket leak can still be reproduced in the Route Guide example by breaking out of the loop in the _ListFeatures()_ method before reading all the responses from the server.

I can also confirm that explicitly cancelling the call does mean that the socket is closed correctly - **however** there still needs to be one _MoveNext()_ call after the cancellation.  Without the _MoveNext()_ there is still a socket leak.  The call to _MoveNext()_ after the cancellation does results in an exception being thrown with _StatusCode.Cancelled_ that needs to be handled.

Here is the altered code in the ListFeatures() client method:

```c#
// create a token so that we can later cancel the call
var tokenSource = new CancellationTokenSource();

using (var call = client.ListFeatures(request, new CallOptions(cancellationToken: tokenSource.Token)))
{
	var responseStream = call.ResponseStream;
	StringBuilder responseLog = new StringBuilder("Result: ");

	while (await responseStream.MoveNext())
	{
		Feature feature = responseStream.Current;
		responseLog.Append(feature.ToString());
		//  Break out of loop to reproduce #8451
		Log("Exiting on first iteration to repro bug #8451");
		
		// Cancel the call
		tokenSource.Cancel();
		// Need to do an explicit MoveNext to get the Cancelled status otherwise
		// the socket is left open
		try
		{
			await responseStream.MoveNext();
		}
		catch (RpcException e) when (e.Status.StatusCode == StatusCode.Cancelled)
		{
			Log("Call cancelled by the client!");
		}

		break;
	}
	Log(responseLog.ToString());
}
```

It would still be nice if the Dispose() of the call did the cancel if necessary without it having to be explicitly done in the code. It looks like the behaviour would be different in the **gRPC for .NET** implementation but I've not tried it to confirm it - the Microsoft documentation says:

>A gRPC call can be canceled in the client by passing a cancellation token with CallOptions.CancellationToken **or calling Dispose** on the call.

https://docs.microsoft.com/en-us/aspnet/core/grpc/deadlines-cancellation?view=aspnetcore-6.0

**Comment by kkm000** [comment]
@tonydnewell, I'm just thinking aloud; not sure at all that this is doable. The `CancellationTokenSource` supports a callback, immediately invoked upon a call to `.Cancel` of it's "token" (internally, token source and token are a single entity, and the source itself and the token are logical interfaces on it, designed with a separation of concerns in mind). Can we somehow hack into this mechanism to make it perform the final try/catch in your example automagically? I.e. call `MoveNext, catch and silently ignore the exception with the same `when (e.Status.StatusCode == StatusCode.Cancelled)` clause, while letting other exception codes to fall through?

/cc @jtattermusch

**Comment by tonydnewell** [comment]
@kkm000 I've investigated moving the cleanup logic to a cancellation callback but it appears the callback is invoked too early in the cancellation process - the exception is never thrown (MoveNext succeeds) and the socket still leaks.

**Comment by kkm000** [comment]
Thanks for checking @tonydnewell! Too bad my wild idea didn't work out... :(

**Comment by jtattermusch** [comment]
@tonydnewell  The fact that call.MoveNext() is still required after requesting call cancellation (regardless whether it's through cancellationToken or call.Dispose()) seems odd to me. It would be good to investigate a bit more why that is the case - and subsequently fix it.
For calls that don't get cancelled, there is little we can do about not requiring the MoveNext() calls - this is simply a side effect of how the C# API has been designed and of how the C-core native component works.  Requiring the user to drain the response stream for non-cancelled calls makes sense IMHO, but it would be nice if this wasn't strictly required for cancelled calls.

One thing to check:
- if you sleep for a few seconds after call.Dispose(), does the call still leak? This could also be a "race" between the call's cancellation taking place and the channel being shutdown.

**Comment by tonydnewell** [comment]
A delay does not make a difference - it is not a race condition.

The problem is that _readingDone_ is _false_ in _AsyncCallBase.ReleaseResourcesIfPossible()_ therefore the resources are not released (_Dispose_ is not called on the native call).  If I set _readingDone_ to _true_ if _cancelRequested_ is _true_ in _ReleaseResourcesIfPossible()_ then the resources are released and there is no socket leak.  I don't think it is a problem to say reading is done (and maybe loosing some date not yet read) as the client has already decided they are giving up with the call.

However I've not yet determined whether this is the correctly place to set this - maybe it should be in _AsyncCall.HandleFinished()_ instead, or if there are other consequences, or whether other types of call are affected. I will investigate further.

Note: _cancelRequested_ is true for both an explicit cancel and also when _Dispose_ is called on the call in the client - so an explicit cancel is not needed

**Comment by jtattermusch** [comment]
- I think we should double check that the doc-comments are clear about the situations in which it's required to drain the call.responseStream (currently it's "always") and update the docs if not.

- we can try to come up with a partial fix where requesting the call cancellation (either through triggering the cancellationToken or call.Dispose()) wouldn't remove the need to drain the call.ResponseStream. This needs to be done carefully since it can break call's behavior in other situations.

**Comment by tonydnewell** [comment]
@jtattermusch can close - the fix is in release v1.46.5
